lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepForward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
+
+    stepForward: ->
+      @tokenIterator.stepForward()
+      curToken = @tokenIterator.getCurrentToken()
+      {row: tokenRow, column: tokenColumn} = @tokenIterator.getCurrentTokenPosition()
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepBackward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
+
+
+  class EquationRangeHandler
+    @BEGIN_EQUATION_TOKEN_SEQUENCE: [
+      { type: "storage.type", value: "\\begin" }
+      { type: "lparen", value: "{" }
+      { type: "variable.parameter", value: "equation" }
+      { type: "rparen", value: "}" }
+    ]
+    @END_EQUATION_TOKEN_SEQUENCE: [
+      { type: "storage.type", value: "\\end" }
+      { type: "lparen", value: "{" }
+      { type: "variable.parameter", value: "equation" }
+      { type: "rparen", value: "}" }
+    ]
+
+    constructor: (@editor) ->
+
+    equalTokens: (token1, token2) ->
+      if token1? and token2?
+        return token1.type == token2.type and token1.value == token2.value
+      else return if token1? or token2? then false else true
+
+    getEquationStart: (tokenIterator) ->
+      # following cycle pushes tokenIterator to the end of
+      # beginning sequence, if it is inside one
+      for token in @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE
+        if @equalTokens(token, tokenIterator.getCurrentToken())
+          tokenIterator.stepForward()
+      curSequenceIndex = @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+      curEquationStart = null
+      while curSequenceIndex >= 0
+        if @equalTokens(
+            @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
+            tokenIterator.stepBackward())
+          if curSequenceIndex == @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+            curTokenPosition = tokenIterator.getCurrentTokenPosition()
+            curEquationStart = {
+              row: curTokenPosition.row
+              column: curTokenPosition.column + tokenIterator.getCurrentToken().value.length
+            }
+          curSequenceIndex -= 1
+        else
+          curSequenceIndex = @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+          curEquationStart = null
+      return curEquationStart
+
+    getEquationEnd: (tokenIterator) ->
+      # following cycle pushes tokenIterator to the start of
+      # ending sequence, if it is inside one
+      for token in @constructor.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
+        if @equalTokens(token, tokenIterator.getCurrentToken())
```

Comment:
any sense to leave `equalTokens` as standalone function? This will save us some CPU cycles and it may matter because it is called frequently